record/record_test.go

```diff
@@ -0,0 +1,240 @@
+		})
+	})
+
+	Describe("Timestamp handling", func() {
+
+		inputTimestampToExpectedOutput := map[interface{}]int64{
+			// Modern Fluent Bit does uses FLBTime
+			output.FLBTime{time.Unix(1234567890, 123456789)}: 1234567890123,
+
+			// We've seen older of Fluent Bit versions use uint64
+			// (generally being sent in seconds, but we handle other granularities out of paranoia)
+			uint64(1234567890):          1234567890000,
+			uint64(1234567890123):       1234567890123,
+			uint64(1234567890123456):    1234567890123,
+			uint64(1234567890123456789): 1234567890123,
+		}
+
+		for inputTimestamp, expectedOutputTime := range inputTimestampToExpectedOutput {
+			// Lock in current values (otherwise all tests will run with the last values in the map)
+			input := inputTimestamp
+			expected := expectedOutputTime
+
+			It("handles timestamps of various types and granularites : "+fmt.Sprintf("%v", input),
+				func() {
+					inputMap := make(FluentBitRecord)
+
+					foundOutput := RemapRecord(inputMap, input, pluginVersion)
+
+					Expect(foundOutput["timestamp"]).To(Equal(expected))
+				},
+			)
+		}
+
+		It("ignores timestamps of unhandled types", func() {
+			inputMap := make(FluentBitRecord)
+
+			// We don't handle string types
+			foundOutput := RemapRecord(inputMap, "1234567890", pluginVersion)
+
+			Expect(foundOutput["timestamp"]).To(BeNil())
+		})
+	})
+
+	Describe("Record packaging", func() {
+
+		It("returns an empty array of packages if the provided slice is nil", func() {
+			// When
+			packagedRecords, err := PackageRecords(nil)
+
+			// Then
+			Expect(err).To(BeNil())
+			Expect(packagedRecords).To(Not(BeNil()))
+			Expect(packagedRecords).To(HaveLen(0))
+		})
+
+		It("returns an empty array of packages if the provided slice is empty", func() {
+			// When
+			packagedRecords, err := PackageRecords([]LogRecord{})
+
+			// Then
+			Expect(err).To(BeNil())
+			Expect(packagedRecords).To(Not(BeNil()))
+			Expect(packagedRecords).To(HaveLen(0))
+		})
+
+		It("returns a compressed JSON array when an array of records is provided", func() {
+			// Given
+			logRecords := []LogRecord{
+				{
+					"timestamp": 1,
+					"message":   "Some message 1",
+				},
+				{
+					"timestamp": 2,
+					"message":   "Some message 2",
+				},
+			}
+			expectedJson := `[{"message":"Some message 1","timestamp":1},{"message":"Some message 2","timestamp":2}]`
+
+			// When
+			packagedRecords, err := PackageRecords(logRecords)
+
+			// Then
+			Expect(err).To(BeNil())
+			Expect(packagedRecords).To(Not(BeNil()))
+			// The two records are compressed into a single byte buffer, as they don't exceed 1MB
+			Expect(packagedRecords).To(HaveLen(1))
+
+			// Decompress and validate record
+			uncompressedJson, err := uncompressRecord(packagedRecords[0])
+			if err != nil {
+				Fail("Could not uncompress record 0")
+			}
+			Expect(uncompressedJson).To(Equal(expectedJson))
+		})
+
+		It("discards a log record if its compressed size exceeds 1MB in size", func() {
```

Comment:
Is there a test where log records is broken up into multiple (maybe > 4) batches? Like not just that one record is too big, but where all of them compressed are too big

Reply:
It took me a while but... Done! :-)